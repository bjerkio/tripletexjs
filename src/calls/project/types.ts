import { DefaultTripletexInputs } from '../../types';

export interface ListProjectsInput extends DefaultTripletexInputs {
  /**
   * List of IDs
   */
  id?: string[];

  /**
   * Containing
   */
  name?: string;

  /**
   * Equals
   */
  number?: string;

  isOffer?: boolean;

  /**
   * List of IDs
   */
  projectManagerId?: string[];

  /**
   * List of IDs
   */
  employeeInProjectId?: string[];

  /**
   * List of IDs
   */
  departmentId?: string[];

  startDateFrom?: Date;
  startDateTo?: Date;

  endDateFrom?: Date;
  endDateTo?: Date;

  isClosed?: boolean;

  /**
   * Equals
   */
  customerId?: string;

  /**
   * Containing
   */
  externalAccountsNumber?: string;
}
