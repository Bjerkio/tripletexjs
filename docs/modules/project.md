[tripletexjs](../README.md) › [Project](project.md)

# Namespace: Project

## Index

### Enumerations

* [displayNameFormat](../enums/project.displaynameformat.md)

### Properties

* [changes](project.md#optional-readonly-changes)
* [contact](project.md#optional-readonly-contact)
* [contributionMarginPercent](project.md#optional-readonly-contributionmarginpercent)
* [currency](project.md#optional-currency)
* [customer](project.md#optional-customer)
* [deliveryAddress](project.md#optional-readonly-deliveryaddress)
* [department](project.md#optional-department)
* [description](project.md#optional-description)
* [discountPercentage](project.md#optional-readonly-discountpercentage)
* [displayName](project.md#optional-readonly-displayname)
* [endDate](project.md#optional-enddate)
* [externalAccountsNumber](project.md#optional-externalaccountsnumber)
* [fixedprice](project.md#optional-fixedprice)
* [forParticipantsOnly](project.md#optional-forparticipantsonly)
* [generalProjectActivitiesPerProjectOnly](project.md#optional-generalprojectactivitiesperprojectonly)
* [hierarchyNameAndNumber](project.md#optional-readonly-hierarchynameandnumber)
* [id](project.md#optional-id)
* [invoicingPlan](project.md#optional-readonly-invoicingplan)
* [isClosed](project.md#optional-isclosed)
* [isFixedPrice](project.md#optional-isfixedprice)
* [isInternal](project.md#isinternal)
* [isOffer](project.md#optional-readonly-isoffer)
* [isPriceCeiling](project.md#optional-ispriceceiling)
* [isReadyForInvoicing](project.md#optional-isreadyforinvoicing)
* [mainProject](project.md#optional-mainproject)
* [markUpFeesEarned](project.md#optional-markupfeesearned)
* [markUpOrderLines](project.md#optional-markuporderlines)
* [name](project.md#name)
* [number](project.md#optional-number)
* [numberOfProjectParticipants](project.md#optional-readonly-numberofprojectparticipants)
* [numberOfSubProjects](project.md#optional-readonly-numberofsubprojects)
* [orderLines](project.md#optional-readonly-orderlines)
* [participants](project.md#optional-participants)
* [preliminaryInvoice](project.md#optional-readonly-preliminaryinvoice)
* [priceCeilingAmount](project.md#optional-priceceilingamount)
* [projectActivities](project.md#optional-projectactivities)
* [projectCategory](project.md#optional-projectcategory)
* [projectHourlyRates](project.md#optional-projecthourlyrates)
* [projectManager](project.md#projectmanager)
* [startDate](project.md#startdate)
* [url](project.md#optional-readonly-url)
* [vatType](project.md#optional-vattype)
* [version](project.md#optional-version)

## Properties

### `Optional` `Readonly` changes

• **changes**? : *Array‹[Change](change.md)›*

___

### `Optional` `Readonly` contact

• **contact**? : *[Contact](../interfaces/contact.md)*

___

### `Optional` `Readonly` contributionMarginPercent

• **contributionMarginPercent**? : *undefined | number*

___

### `Optional` currency

• **currency**? : *[Currency](../interfaces/currency.md)*

___

### `Optional` customer

• **customer**? : *[Customer](customer.md)*

The project's customer

___

### `Optional` `Readonly` deliveryAddress

• **deliveryAddress**? : *[DeliveryAddress](../interfaces/deliveryaddress.md)*

___

### `Optional` department

• **department**? : *[Department](../interfaces/department.md)*

___

### `Optional` description

• **description**? : *undefined | string*

___

### `Optional` `Readonly` discountPercentage

• **discountPercentage**? : *undefined | number*

Project discount percentage.

___

### `Optional` `Readonly` displayName

• **displayName**? : *undefined | string*

___

### `Optional` endDate

• **endDate**? : *undefined | string*

___

### `Optional` externalAccountsNumber

• **externalAccountsNumber**? : *undefined | string*

___

### `Optional` fixedprice

• **fixedprice**? : *undefined | number*

Fixed price amount, in the project's currency.

___

### `Optional` forParticipantsOnly

• **forParticipantsOnly**? : *undefined | false | true*

Set to true if only project participants can register information on the project

___

### `Optional` generalProjectActivitiesPerProjectOnly

• **generalProjectActivitiesPerProjectOnly**? : *undefined | false | true*

Set to true if a general project activity must be linked to project to allow time tracking.

___

### `Optional` `Readonly` hierarchyNameAndNumber

• **hierarchyNameAndNumber**? : *undefined | string*

___

### `Optional` id

• **id**? : *undefined | number*

___

### `Optional` `Readonly` invoicingPlan

• **invoicingPlan**? : *Array‹[Invoice](invoice.md)›*

Invoicing plans tied to the project

___

### `Optional` isClosed

• **isClosed**? : *undefined | false | true*

___

### `Optional` isFixedPrice

• **isFixedPrice**? : *undefined | false | true*

Project is fixed price if set to true, hourly rate if set to false.

___

###  isInternal

• **isInternal**: *boolean*

___

### `Optional` `Readonly` isOffer

• **isOffer**? : *undefined | false | true*

___

### `Optional` isPriceCeiling

• **isPriceCeiling**? : *undefined | false | true*

Set to true if an hourly rate project has a price ceiling.

___

### `Optional` isReadyForInvoicing

• **isReadyForInvoicing**? : *undefined | false | true*

___

### `Optional` mainProject

• **mainProject**? : *[Project](project.md)*

___

### `Optional` markUpFeesEarned

• **markUpFeesEarned**? : *undefined | number*

Set mark-up (%) for fees earned.

___

### `Optional` markUpOrderLines

• **markUpOrderLines**? : *undefined | number*

Set mark-up (%) for order lines.

___

###  name

• **name**: *string*

___

### `Optional` number

• **number**? : *undefined | string*

___

### `Optional` `Readonly` numberOfProjectParticipants

• **numberOfProjectParticipants**? : *undefined | number*

___

### `Optional` `Readonly` numberOfSubProjects

• **numberOfSubProjects**? : *undefined | number*

___

### `Optional` `Readonly` orderLines

• **orderLines**? : *Array‹[ProjectOrderLine](../interfaces/projectorderline.md)›*

Order lines tied to the order

___

### `Optional` participants

• **participants**? : *Array‹[ProjectParticipant](../interfaces/projectparticipant.md)›*

Link to individual project participants.

___

### `Optional` `Readonly` preliminaryInvoice

• **preliminaryInvoice**? : *[Invoice](invoice.md)*

___

### `Optional` priceCeilingAmount

• **priceCeilingAmount**? : *undefined | number*

Price ceiling amount, in the project's currency.

___

### `Optional` projectActivities

• **projectActivities**? : *Array‹[ProjectActivity](../interfaces/projectactivity.md)›*

Project Activities

___

### `Optional` projectCategory

• **projectCategory**? : *[ProjectCategory](../interfaces/projectcategory.md)*

___

### `Optional` projectHourlyRates

• **projectHourlyRates**? : *Array‹[ProjectHourlyRate](projecthourlyrate.md)›*

Project Rate Types tied to the project.

___

###  projectManager

• **projectManager**: *[Employee](employee.md)*

___

###  startDate

• **startDate**: *string*

___

### `Optional` `Readonly` url

• **url**? : *undefined | string*

___

### `Optional` vatType

• **vatType**? : *[VatType](../interfaces/vattype.md)*

The default vat type for this project.

___

### `Optional` version

• **version**? : *undefined | number*
