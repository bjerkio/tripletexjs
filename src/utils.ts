import * as rt from 'runtypes';

export function pickFromObject<
  T extends Record<string, unknown>,
  K extends keyof T,
>(subject: T, ...keys: K[]): Pick<T, K> {
  const pairs = keys
    .map(key => [key, subject[key]])
    .filter(([, val]) => val !== undefined)
    .map(([key, val]) => [key, val]);
  return Object.fromEntries(pairs);
}

export function withRuntype<T>(validator: rt.Runtype<T>) {
  return (data: unknown) => {
    return validator.check(data);
  };
}

export function multipleValuesEnvelope<T>(type: rt.Runtype<T>) {
  return rt
    .Record({
      from: rt.Number,
      count: rt.Number,
      versionDigest: rt.String.nullable().optional(),
      values: rt.Array(type).asReadonly(),
    })
    .asReadonly();
}

export function singleValueEnvelope<T>(type: rt.Runtype<T>) {
  return rt
    .Record({
      value: type,
    })
    .asReadonly();
}
