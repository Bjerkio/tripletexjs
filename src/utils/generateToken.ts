import { OpenAPI } from '../generated/core/OpenAPI';
import { TokenSessionService } from '../generated/services/TokenSessionService';
import { invariant } from './invariant';

/**
 * generateToken helps creating a token. It sets token for all requests.
 * 
 * @param consumerToken Token of the API consumer
 * @param employeeToken The employees token
 * @param expirationDate Expiration date for the combined token
 * @example
 * ```typescript
 * const expDate = new Date();
 * expDate.setDate(expDate.getDate() + 1);
 * await generateToken({
 *   consumerToken: '000-000-000',
 *   employeeToken: '100-100-100',
 *   expirationDate: expDate,
 * });
 * ```
 */
export async function generateToken({
  expirationDate,
  ...tokenSessionArgs
}: {
  consumerToken: string;
  employeeToken: string;
  expirationDate: Date;
}) {
  if (OpenAPI.TOKEN) {
    return false;
  }

  const expDate = expirationDate.toISOString().split('T')[0];

  const session = await TokenSessionService.tokenSessionCreateCreate({
    expirationDate: expDate,
    ...tokenSessionArgs,
  });

  invariant(session.value.token, 'Not able to get token');

  OpenAPI.TOKEN = session.value;
}

/**
 * 0 or blank means the company of the employee.
 * Any other value means accountant clients. Use /company/>withLoginAccess to get a list of those.
 * @param username defaults to '0'
 */
export async function setUsername(username: string | (() => Promise<string>)) {
  OpenAPI.USERNAME = username;
}