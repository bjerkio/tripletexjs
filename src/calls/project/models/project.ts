import * as rt from 'runtypes';
import { multipleValuesEnvelope, resourceRef, singleValueEnvelope } from '../../../utils';

const projectRt = rt.Record({
  id: rt.Number,
  version: rt.Number.nullable().optional(),
  name: rt.String,
  displayName: rt.String,
  number: rt.String.nullable().optional(),
  description: rt.String.nullable().optional(),
  projectManager: rt.Record({
    id: rt.Number,
  }),
  department: rt
    .Record({
      id: rt.Number,
    })
    .nullable()
    .optional(),
  mainProject: rt
    .Record({
      id: rt.Number,
    })
    .nullable()
    .optional(),
  startDate: rt.String,
  endDate: rt.String.nullable().optional(),
  customer: resourceRef,
  isClosed: rt.Boolean,
  isReadyForInvoicing: rt.Boolean,
  isInternal: rt.Boolean,
  isFixedPrice: rt.Boolean,
  projectCategory: resourceRef,
  deliveryAddress: resourceRef,
  displayNameFormat: rt.Union(
    rt.Literal('NAME_STANDARD'),
    rt.Literal('NAME_INCL_CUSTOMER_NAME'),
    rt.Literal('NAME_INCL_PARENT_NAME'),
    rt.Literal('NAME_INCL_PARENT_NUMBER'),
    rt.Literal('NAME_INCL_PARENT_NAME_AND_NUMBER'),
  ),
  reference: rt.String.nullable().optional(),
  externalAccountsNumber: rt.String.nullable().optional(),
  vatType: resourceRef,
  fixedprice: rt.Number,
  currency: resourceRef,
  markUpOrderLines: rt.Number,
  markUpFeesEarned: rt.Number,
  isPriceCeiling: rt.Boolean,
  priceCeilingAmount: rt.Number,
  projectHourlyRates: rt.Array(resourceRef),
  forParticipantsOnly: rt.Boolean,
  participants: rt.Array(resourceRef),
  contact: resourceRef,
  attention: resourceRef,
  invoiceComment: rt.String,
  generalProjectActivitiesPerProjectOnly: rt.Boolean,
  projectActivities: rt
    .Array(
      rt.Record({
        id: rt.Number,
      }),
    )
    .nullable()
    .optional(),
});

export type Project = rt.Static<typeof projectRt>;

export const listProjectResponseRt = multipleValuesEnvelope(projectRt);
export const createProjectResponseRt = singleValueEnvelope(projectRt);
