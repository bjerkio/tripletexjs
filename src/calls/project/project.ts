import { serializeQuery, withRuntype } from '../../utils';
import { TripletexBase } from '../base';
import { listProjectResponseRt } from './models/project';
import { ListProjectsInput } from './types';
export * from './models/project';


export class TripletexProject extends TripletexBase {
  list(input?: ListProjectsInput) {
    const call = this.authenticatedCall() //
      .args<{
        input?: ListProjectsInput;
      }>()
      .path('/v2/project')
      .query(args => (args.input ? serializeQuery(args.input) : {}))
      .method('get')
      .parseJson(withRuntype(listProjectResponseRt))
      .build();

    return this.performRequest(sessionToken => call({ input, sessionToken }));
  }
}
