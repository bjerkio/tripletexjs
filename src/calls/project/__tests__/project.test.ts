import nock from 'nock';
import { invariant } from 'ts-invariant';
import { parseRuntypeValidationError } from '../../../utils';
import { TripletexProject } from '../project';

const baseUrl = process.env.BASE_URL || 'https://api.tripletex.io/';
const sessionToken = process.env.SESSION_TOKEN || 'a-session-token';

describe('activity class', () => {
  let client: TripletexProject;

  beforeEach(() => {
    client = new TripletexProject({ baseUrl }, sessionToken);
  });

  it('should list activities', async () => {
    nock('https://api.tripletex.io:443', { encodedQueryParams: true })
      .get('/v2/project')
      .reply(200, {
        fullResultSize: 1,
        from: 0,
        count: 1,
        versionDigest: "'If-None-Match' header not specified",
        values: [
          {
            id: 1560889,
            version: 0,
            url: 'api.tripletex.io/v2/project/1560889',
            name: 'Test',
            number: '1',
            displayName: '1 Test',
            description: '',
            projectManager: {
              id: 1734852,
              url: 'api.tripletex.io/v2/employee/1734852',
            },
            department: {
              id: 88885,
              url: 'api.tripletex.io/v2/department/88885',
            },
            mainProject: null,
            startDate: '2022-04-04',
            endDate: null,
            customer: null,
            isClosed: false,
            isReadyForInvoicing: false,
            isInternal: false,
            isOffer: false,
            isFixedPrice: false,
            projectCategory: null,
            deliveryAddress: null,
            displayNameFormat: 'NAME_STANDARD',
            reference: '',
            externalAccountsNumber: '',
            discountPercentage: 0,
            vatType: { id: 3, url: 'api.tripletex.io/v2/ledger/vatType/3' },
            fixedprice: 0,
            contributionMarginPercent: 0,
            numberOfSubProjects: 0,
            numberOfProjectParticipants: 1,
            orderLines: [],
            currency: { id: 1, url: 'api.tripletex.io/v2/currency/1' },
            markUpOrderLines: 0,
            markUpFeesEarned: 0,
            isPriceCeiling: false,
            priceCeilingAmount: 0,
            projectHourlyRates: [
              {
                id: 50292,
                url: 'api.tripletex.io/v2/project/hourlyRates/50292',
              },
            ],
            forParticipantsOnly: false,
            participants: [
              {
                id: 16800,
                url: 'api.tripletex.io/v2/project/participant/16800',
              },
            ],
            contact: null,
            attention: null,
            invoiceComment: '',
            invoicingPlan: [],
            preliminaryInvoice: null,
            generalProjectActivitiesPerProjectOnly: false,
            projectActivities: [],
            hierarchyNameAndNumber: '1 Test',
            invoiceDueDate: 0,
            invoiceReceiverEmail: '',
            accessType: 'WRITE',
            useProductNetPrice: false,
            ignoreCompanyProductDiscountAgreement: false,
          },
        ],
      });
    const activities = await client.list({});

    parseRuntypeValidationError(activities.error);

    invariant(activities.success);

    expect(activities.body.values).toMatchSnapshot();
  });
});
