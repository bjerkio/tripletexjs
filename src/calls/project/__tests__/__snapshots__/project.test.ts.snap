// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`activity class should list activities 1`] = `
Array [
  Object {
    "accessType": "WRITE",
    "attention": null,
    "contact": null,
    "contributionMarginPercent": 0,
    "currency": Object {
      "id": 1,
      "url": "api.tripletex.io/v2/currency/1",
    },
    "customer": null,
    "deliveryAddress": null,
    "department": Object {
      "id": 88885,
      "url": "api.tripletex.io/v2/department/88885",
    },
    "description": "",
    "discountPercentage": 0,
    "displayName": "1 Test",
    "displayNameFormat": "NAME_STANDARD",
    "endDate": null,
    "externalAccountsNumber": "",
    "fixedprice": 0,
    "forParticipantsOnly": false,
    "generalProjectActivitiesPerProjectOnly": false,
    "hierarchyNameAndNumber": "1 Test",
    "id": 1560889,
    "ignoreCompanyProductDiscountAgreement": false,
    "invoiceComment": "",
    "invoiceDueDate": 0,
    "invoiceReceiverEmail": "",
    "invoicingPlan": Array [],
    "isClosed": false,
    "isFixedPrice": false,
    "isInternal": false,
    "isOffer": false,
    "isPriceCeiling": false,
    "isReadyForInvoicing": false,
    "mainProject": null,
    "markUpFeesEarned": 0,
    "markUpOrderLines": 0,
    "name": "Test",
    "number": "1",
    "numberOfProjectParticipants": 1,
    "numberOfSubProjects": 0,
    "orderLines": Array [],
    "participants": Array [
      Object {
        "id": 16800,
        "url": "api.tripletex.io/v2/project/participant/16800",
      },
    ],
    "preliminaryInvoice": null,
    "priceCeilingAmount": 0,
    "projectActivities": Array [],
    "projectCategory": null,
    "projectHourlyRates": Array [
      Object {
        "id": 50292,
        "url": "api.tripletex.io/v2/project/hourlyRates/50292",
      },
    ],
    "projectManager": Object {
      "id": 1734852,
      "url": "api.tripletex.io/v2/employee/1734852",
    },
    "reference": "",
    "startDate": "2022-04-04",
    "url": "api.tripletex.io/v2/project/1560889",
    "useProductNetPrice": false,
    "vatType": Object {
      "id": 3,
      "url": "api.tripletex.io/v2/ledger/vatType/3",
    },
    "version": 0,
  },
]
`;
